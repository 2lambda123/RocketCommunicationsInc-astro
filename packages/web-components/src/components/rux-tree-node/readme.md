# rux-tree-node

<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description             | Type      | Default |
| ---------- | ---------- | ----------------------- | --------- | ------- |
| `expanded` | `expanded` | Sets the expanded state | `boolean` | `false` |
| `selected` | `selected` | Sets the selected state | `boolean` | `false` |


## Events

| Event                  | Description                                                                            | Type                  |
| ---------------------- | -------------------------------------------------------------------------------------- | --------------------- |
| `ruxtreenodecollapsed` | Fires when the user collapses a tree node and emits the node's id on the event.detail. | `CustomEvent<string>` |
| `ruxtreenodeexpanded`  | Fires when the user expands a tree node and emits the node's id on the event.detail.   | `CustomEvent<string>` |
| `ruxtreenodeselected`  | Fires when the user selects a tree node and emits the node's id on the event.detail.   | `CustomEvent<string>` |


## Methods

### `setExpanded(value: boolean) => Promise<void>`

Sets the expanded state

#### Returns

Type: `Promise<void>`



### `setSelected(value: boolean) => Promise<void>`

Sets the selected state

#### Returns

Type: `Promise<void>`




## Shadow Parts

| Part          | Description |
| ------------- | ----------- |
| `"indicator"` |             |
| `"node"`      |             |
| `"text"`      |             |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
