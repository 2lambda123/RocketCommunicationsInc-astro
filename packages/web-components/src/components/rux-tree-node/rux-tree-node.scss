:host {
    /**
    * @prop --tree-accent-color: tree accent color
    */
    --tree-accent-color: var(--color-interactive-default);

    /**
    * @prop --tree-hover-background-color: tree hover background color
    */
    --tree-hover-background-color: var(--color-interactive-hover);

    /**
    * @prop --tree-hover-text-color: tree hover text color
    */
    --tree-hover-text-color: var(--color-text-inverse);

    /**
    * @prop --tree-selected-border-color: tree selected border color
    */
    --tree-selected-border-color: var(--color-interactive-muted);

    /**
    * @prop --tree-selected-accent-color: tree selected accent color
    */
    --tree-selected-accent-color: var(--color-interactive-default);

    /**
    * @prop --tree-expanded-border-color: tree expanded border color
    */
    --tree-expanded-border-color: var(--color-background-base-default);

    position: relative;
    box-sizing: border-box;
    width: 100%;
    padding: 0;
    margin: 0;
    font-family: var(--font-body-1-font-family);
    font-size: var(--font-body-1-font-size);
    font-weight: var(--font-body-1-font-weight);
    letter-spacing: var(--font-body-1-letter-spacing);
    color: var(--tree-text-color);
    user-select: none;
    display: block;
}

:host([hidden]) {
    display: none;
}

:host([aria-expanded='true'][aria-level='1']:not([aria-selected='true'])) {
    .parent::after {
        border-bottom: solid 1px var(--tree-expanded-border-color);
    }
}

:host([aria-selected='true']) {
    .parent {
        &::after {
            content: '';
            height: 30px;
            width: 100%;
            left: 0;
            z-index: 0;
            position: absolute;
            transition: background-color 0.0967s ease-in 0s;
            border-top: 1px solid var(--tree-selected-border-color);
            border-bottom: 1px solid var(--tree-selected-border-color);
        }
    }
}

:host([aria-level='1']) {
    .parent {
        font-weight: bold;
        padding-left: 0.5rem;
        &::after {
            content: '';
            height: 31px;
            width: 100%;
            left: 0;
            z-index: 0;
            position: absolute;
            transition: background-color 0.0967s ease-in 0s;
        }
    }
    .tree-node:not(.tree-node--has-children) {
        .parent {
            padding-left: 2rem;
        }
    }
}

:host([aria-level='2']) {
    .tree-node:not(.tree-node--has-children) {
        .parent {
            padding-left: 3.5rem;
        }
    }
}

:host([aria-level='3']) {
    .tree-node:not(.tree-node--has-children) {
        .parent {
            padding-left: 5rem;
        }
    }
    .parent {
        padding-left: 3.5rem;
    }
}

:host([aria-level='4']) {
    .tree-node:not(.tree-node--has-children) {
        .parent {
            padding-left: 6.5rem;
        }
    }
    .parent {
        padding-left: 5rem;
    }
}

.parent {
    height: 2rem;
    padding: 0 0.5rem 0 2rem;
    display: flex;
    align-items: center;
    &:hover {
        color: var(--tree-hover-text-color);
        background: var(--tree-hover-background-color);
    }
}

.children {
    display: none;
}

.tree-node--expanded {
    > .children {
        display: block !important;
    }
    .arrow {
        transform: rotate(90deg);
    }
}

.arrow {
    position: relative;
    cursor: pointer;
    width: 0.35rem;
    margin-right: 1rem;
    margin-left: 0.15rem;
    background-color: transparent;
    transition: transform 0.167s ease-in-out 0s;
    z-index: 11;
    &::before {
        content: '';
        display: block;
        height: 1.5rem;
        width: 1.5rem;
        top: -0.15rem;
        left: -0.65rem;
        position: absolute;
    }
    &::after {
        content: '';
        width: 0;
        height: 0;
        border-style: solid;
        border-width: 0.35rem 0 0.35rem 0.35rem;
        border-color: transparent transparent transparent
            var(--tree-accent-color);
        display: inline-block;
    }
}

::slotted(rux-status) {
    margin: 0 0.625rem;
}

::slotted(rux-tree-node) {
    position: relative;
    &::before {
        content: '';
        order: 1;
        position: absolute;
        top: 0;
        cursor: pointer;
        width: 20px;
        height: 20px;
        margin-right: 1rem;
        margin-left: 0.15rem;
        left: -40px;
        background-color: transparent;
        transition: transform 0.167s ease-in-out 0s;
        z-index: 11;
    }
}
