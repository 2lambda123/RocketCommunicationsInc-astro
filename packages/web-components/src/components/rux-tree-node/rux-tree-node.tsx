/* eslint react/jsx-no-bind: 0 */ // --> OFF
import {
    Prop,
    Event,
    EventEmitter,
    State,
    Element,
    Component,
    Host,
    Listen,
    Method,
    Watch,
    h,
} from '@stencil/core'

let id = 0
@Component({
    tag: 'rux-tree-node',
    styleUrl: 'rux-tree-node.scss',
    shadow: true,
})

/**
 * @slot - The parent node content
 * @slot node - Renders a child node within the current node
 */
export class RuxTreeNode {
    private componentId = `node-${++id}`
    @Element() el!: HTMLRuxTreeNodeElement
    @State() children: Array<HTMLRuxTreeNodeElement> = []

    /**
     * Sets the expanded state
     */
    @Prop({ mutable: true, reflect: true }) expanded = false
    /**
     * Sets the selected state
     */
    @Prop({ mutable: true, reflect: true }) selected = false

    /**
     * Emit when user selects a tree node
     */
    @Event({ eventName: 'ruxtreenodeselected' })
    ruxTreeNodeSelected!: EventEmitter<string>

    @Watch('expanded')
    handleExpandedChange(newValue: boolean) {
        this.setExpanded(newValue)
    }

    @Watch('selected')
    handleSelectedChange(newValue: boolean) {
        this.setSelected(newValue)
    }

    @Listen('keydown', { passive: true })
    handleKeyDown(ev: KeyboardEvent) {
        if (ev.target !== ev.currentTarget) {
            return true
        }

        switch (ev.key) {
            case 'ArrowUp':
                ev.preventDefault()
                this._focusNext(-1)
                break
            case 'ArrowRight':
                ev.preventDefault()
                this._expandNextNode()
                break
            case 'ArrowDown':
                ev.preventDefault()
                this._focusNext(1)
                break
            case 'ArrowLeft':
                ev.preventDefault()
                this._collapseParent()
                break
            case 'Enter':
                ev.preventDefault()
                this.setSelected(true)
                break
        }
    }

    connectedCallback() {
        this._handleSlotChange = this._handleSlotChange.bind(this)
    }

    componentWillLoad() {
        this._handleSlotChange()
    }

    get _hasChildren() {
        return this.children.length > 0
    }

    /**
     * Sets the expanded state
     * @param value
     */
    @Method()
    async setExpanded(value: boolean) {
        this.expanded = value
    }

    /**
     * Sets the selected state
     * @param value
     */
    @Method()
    async setSelected(value: boolean) {
        this.selected = value
        if (value) {
            this.ruxTreeNodeSelected.emit(this.componentId)
        }
    }

    private _handleSlotChange() {
        const children = Array.from(
            this.el.querySelectorAll(`[slot="node"]`)
        ) as HTMLRuxTreeNodeElement[]
        this.children = children
        this._setAriaLevel()
    }

    /**
     * Manually set the aria-level attribute.
     * Tree is responsible for setting the root node levels.
     */
    private _setAriaLevel() {
        const level = this.el.getAttribute('aria-level')
        if (level) {
            this.children.map((child) => {
                child.setAttribute('aria-level', `${+level + 1}`)
            })
        }
    }

    private _handleArrowClick(e: MouseEvent) {
        e.stopPropagation()
        this.setExpanded(!this.expanded)
    }

    private _handleTreeNodeClick(e: MouseEvent) {
        e.stopPropagation()
        this.selected = !this.selected
    }

    private _expandNextNode() {
        if (!this.expanded && this._hasChildren) {
            this.setExpanded(true)
        }
    }

    private _focusItem(el: HTMLRuxTreeNodeElement) {
        const parent = el?.shadowRoot?.querySelector('.parent') as HTMLElement
        if (parent) {
            parent.focus()
        }
    }

    private _collapseParent() {
        if (this.expanded) {
            this.setExpanded(false)
        } else if (this.el.parentElement) {
            const parentTreeItemNode:
                | Element
                | null
                | undefined = this.el.parentElement!.closest(
                "[role='treeitem']"
            )

            if (parentTreeItemNode) {
                this._focusItem(parentTreeItemNode as HTMLRuxTreeNodeElement)
            }
        }
    }

    private _focusNext(direction: number) {
        const visibleNodes = this._getVisibleNodes()
        const currentIndex: number = visibleNodes.indexOf(this.el)
        console.log('start focus')
        if (currentIndex !== -1) {
            let nextElement: HTMLRuxTreeNodeElement | undefined =
                visibleNodes[currentIndex + direction]
            if (nextElement !== undefined) {
                // Skips any disabled nodes
                while (nextElement.hasAttribute('disabled')) {
                    const offset: number = direction >= 0 ? 1 : -1
                    nextElement =
                        visibleNodes[currentIndex + direction + offset]
                    if (nextElement) {
                        break
                    }
                }
            }

            if (nextElement !== null) {
                this._focusItem(nextElement as HTMLRuxTreeNodeElement)
            }
        }
    }

    private _getVisibleNodes() {
        const rootTree = this.el.closest("[role='tree']") as HTMLRuxTreeElement
        const nodes = Array.from(rootTree.querySelectorAll('rux-tree-node'))
        return nodes.filter(
            (node: HTMLRuxTreeNodeElement) => node.offsetParent !== null
        )
    }

    render() {
        const attrs = this._hasChildren && { role: 'group' }

        return (
            <Host
                role="treeitem"
                aria-expanded={this.expanded ? 'true' : 'false'}
                aria-selected={this.selected ? 'true' : 'false'}
                onClick={(event: MouseEvent) =>
                    this._handleTreeNodeClick(event)
                }
            >
                <div
                    id={this.componentId}
                    class={{
                        'tree-node': true,
                        'tree-node--expanded': this.expanded,
                        'tree-node--has-children': this._hasChildren,
                        'tree-node--selected': this.selected,
                    }}
                >
                    <div class="parent" tabindex="0">
                        {this._hasChildren && (
                            <i
                                onClick={(e) => this._handleArrowClick(e)}
                                class="arrow"
                            ></i>
                        )}
                        <slot onSlotchange={this._handleSlotChange}></slot>
                    </div>
                    <div {...attrs} class="children">
                        <slot
                            name="node"
                            onSlotchange={this._handleSlotChange}
                        ></slot>
                    </div>
                </div>
            </Host>
        )
    }
}

/*
                    {this.selected && (
                        <div class="selected"></div>
                        )
                    }
*/
