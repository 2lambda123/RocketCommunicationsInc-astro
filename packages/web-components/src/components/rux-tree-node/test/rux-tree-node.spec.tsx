import { newSpecPage } from '@stencil/core/testing'
import { RuxTreeNode } from '../rux-tree-node'

describe('rux-tree-node', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxTreeNode],
            html: `<rux-tree-node></rux-tree-node>`,
        })
        expect(page.root).toMatchSnapshot()
    })

    it('renders children', async () => {
        const page = await newSpecPage({
            components: [RuxTreeNode],
            html: `<rux-tree-node>
            Parent
            <rux-tree-node slot="node">Child</rux-tree-node>
          </rux-tree-node>`,
        })
        expect(page.root).toMatchSnapshot()
    })
})
