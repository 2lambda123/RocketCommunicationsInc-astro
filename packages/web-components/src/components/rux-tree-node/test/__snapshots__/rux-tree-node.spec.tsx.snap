// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-tree-node renders 1`] = `
<rux-tree-node aria-expanded="false" aria-selected="false" role="treeitem">
  <mock:shadow-root>
    <div class="tree-node" id="node-1">
      <div class="parent" tabindex="0">
        <slot></slot>
      </div>
      <div class="children">
        <slot name="node"></slot>
      </div>
    </div>
  </mock:shadow-root>
</rux-tree-node>
`;

exports[`rux-tree-node renders children 1`] = `
<rux-tree-node aria-expanded="false" aria-selected="false" role="treeitem">
  <mock:shadow-root>
    <div class="tree-node tree-node--has-children" id="node-2">
      <div class="parent" tabindex="0">
        <i class="arrow"></i>
        <slot></slot>
      </div>
      <div class="children" role="group">
        <slot name="node"></slot>
      </div>
    </div>
  </mock:shadow-root>
  Parent
  <rux-tree-node aria-expanded="false" aria-selected="false" role="treeitem" slot="node">
    <mock:shadow-root>
      <div class="tree-node" id="node-3">
        <div class="parent" tabindex="0">
          <slot></slot>
        </div>
        <div class="children">
          <slot name="node"></slot>
        </div>
      </div>
    </mock:shadow-root>
    Child
  </rux-tree-node>
</rux-tree-node>
`;
